core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -198,26 +204,30 @@ public void close()
      * transitioning of the object will not work properly.
      */
     public static SourceScheduler newSourcePartitionedSchedulerAsSourceScheduler(
-            SqlStageExecution stage,
+            PipelinedStageExecution stageExecution,
             PlanNodeId partitionedNode,
             SplitSource splitSource,
             SplitPlacementPolicy splitPlacementPolicy,
             int splitBatchSize,
             boolean groupedExecution,
             DynamicFilterService dynamicFilterService,
             TableExecuteContextManager tableExecuteContextManager,
-            BooleanSupplier anySourceTaskBlocked)
+            BooleanSupplier anySourceTaskBlocked,
+            AtomicInteger nextPartitionId,
```

Comment:
You may want to expose this as something other than an AtomicInteger. An AtomicInteger can be modified in ways that could mess up the sequence (e.g., if someone calls `set()`, or forget to increment).

It should be some kind of IdAllocator that could be backed by an AtomicInteger or just a plan integer with proper synchronization.

Reply:
Added `PartitionIdAllocator`

Reply:
Since it is used by the scheduler that is designed to be called by a single thread synchronization shouldn't be needed.